src/EfuncMsg.chpl

```diff
@@ -742,6 +743,37 @@ module EfuncMsg
         return new MsgTuple(repMsg, MsgType.NORMAL); 
     }
 
+    proc efuncArrMsg(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab): MsgTuple throws {
+        var n = msgArgs.get("length").getIntValue();
+        var s = msgArgs.get("size").getIntValue();
+        var namesList = msgArgs.get("nameslist").getList(n);
+        var typesList = msgArgs.get("typeslist").getList(n);
+        var (size, hasStr, names, types) = validateArraysSameLength(n, namesList, typesList, st);
+
+        // Call hashArrays on list of given array names
+        var hashes = hashArrays(size, names, types, st);
+
+        // Init upper and lower 64-bit entries
+        var upper = new shared SymEntry(s, uint);
+        var lower = new shared SymEntry(s, uint);
+
+        // Assign upper and lower bit values to their respective entries
+        forall (i, (up, low)) in zip(0..#s, hashes) {
+            upper.a[i] = up;
+            lower.a[i] = low;
+        }
+
+        var upperName = st.nextName();
+        st.addEntry(upperName, upper);
+        var lowerName = st.nextName();
+        st.addEntry(lowerName, lower);
```

Comment:
This has some added performance benefits by creating the SymEntry from the dist array (eliminates one array creation) and avoiding accessing the `.a` component of the SymEntry during every iteration of the forall loop.

Reply:
I have wanted something like that for `radixSortLSD` calls in places like `uniqueSortWithInverse`. I don't actually think there's a direct way to do it because the memory layout is different (array of 16-byte values vs 2 arrays of 8-byte values), but it would be nice to have a array-of-tuples to tuple-of-arrays function that didn't add much overhead.

In the meantime though, I think @pierce314159's suggestion is the most efficient.


---

In case it's helpful, here's an explanation of how a few different idioms map down in a distributed context:


The current code:
```chpl
forall (i, (up, low)) in zip(0..#s, hashes) {
    upper.a[i] = up;
    lower.a[i] = low;
}
```

actually turns into a parallel loop that only executes on locale 0 because ranges aren't distributed. So a task per core on locale 0 will be executing, but the rest of the machine will be unused. Then for elements that live on locale 1 for example it's actually something like:

```chpl
var tmp_up = GET(up);
var tmp_upper = GET(upper.a);
PUT(tmp_upper[i], tmp_up);
```

So you have communication to GET `up` since it won't always be local, communication to access `upper.a` because upper is a class on locale 0 and `.a` is the distributed field, and then a PUT to store the result into `a[i]`

Slightly improved using distributed domain:

```chpl
forall (i, (up, low)) in zip(upper.domain, hashes) {
    upper.a[i] = up;
    lower.a[i] = low;
}
```

Then up/low will always be local, but you still the class wrapper living on locale 0 so there's comm for getting `upper.a` and `lower.a` each time. You could rewrite this like:

```chpl
ref ua = upper.a;
ref la = lower.a;
forall (i, (up, low)) in zip(upper.domain, hashes) {
    ua[i] = up;
    la[i] = low;
}
```

which is an idiom you've probably seen in other arkouda code. This direct indexing will be optimized such that we don't have to do locality checks for `ua[i]`. In general we'd have to do more complicated indexing logic to figure out where that index lives, but we'll do a runtime check to see that the domains are the same and take a fast path that just accesses local things.

And the final version Pierce suggested:

```chpl
forall (up, low, h) in zip(upper, lower, hashes) {
    (up, low) = h;
}
```

is IMO cleaner, avoids the runtime check for equal domains, and avoids the need to take references. It might be an interesting experiment to try out these various idioms in a standalone program using [CommDiagnostics](https://chapel-lang.org/docs/modules/standard/CommDiagnostics.html) just to get a sense for how things map down. Something like:

```chpl
use CommDiagnostics;
startVerboseComm();
forall (i, (up, low)) in zip(0..#s, hashes) {
    upper.a[i] = up;
    lower.a[i] = low;
}
stopVerboseComm();
writeln(getCommDiagnostics()); // or printCommDiagnosticsTable();
```

Reply:
Wow! Thanks for all the great info @ronawho!!

Reply:
I updated this to use Pierce's version. Thanks for the run down on how these different options work @ronawho !! I didn't realize ranges weren't distributed, kind of assumed they just worked like arrays.